luigi/contrib/beam_dataflow.py

```diff
@@ -0,0 +1,451 @@
+        """
+        If one/some of the input target files are not in the pattern of part-*,
+        we can add the key of the required target and the correct file pattern
+        that should be appended in the command line here. If the input target key is not found
+        in this dict, the file pattern will be assumed to be part-* for that target.
+
+        :return A dictionary of overridden file pattern that is not part-* for the inputs
+        """
+        return {}
+
+    def on_output_validation(self):
+        """
+        Callback that gets called after the Dataflow job has finished
+        successfully if validate_output returns True.
+        """
+        pass
+
+    def cleanup_on_error(self):
+        """
+        Callback that gets called after the Dataflow job has finished
+        unsuccessfully, or validate_output returns False.
+        """
+        pass
+
+    def run(self):
+        cmd_line = self._mk_cmd_line()
+        logger.info(' '.join(cmd_line))
+
+        self.before_run()
+
+        try:
+            self.cmd_line_runner.run(cmd_line, self)
+        except subprocess.CalledProcessError as e:
+            logger.error(e, exc_info=True)
+            self.cleanup_on_error()
+            """
+            Exit Luigi with the same exit code as the Dataflow job process, so
+            users can easily exit the job with code 50 to avoid Styx retries
+            https://github.com/spotify/styx/blob/master/doc/design-overview.md#workflow-state-graph
+            """
+            os._exit(e.returncode)
+
+        self.on_successful_run()
+
+        if self.validate_output():
+            self.on_output_validation()
+        else:
+            self.cleanup_on_error()
+            raise ValueError("Output validation failed")
+
+    def _mk_cmd_line(self):
+        cmd_line = self.dataflow_executable()
+
+        cmd_line.extend(self._get_dataflow_args())
+        cmd_line.extend(self.args())
+        cmd_line.extend(self._format_input_args())
+        cmd_line.extend(self._format_output_args())
+        return cmd_line
+
+    def _get_runner(self):
+        if self.runner in [
+            "DataflowRunner",
+            "DirectRunner"
+        ]:
+            return self.runner
+
+        elif self.runner in [
+            "InProcessPipelineRunner",
+            "BlockingDataflowPipelineRunner"
+        ]:
+            logger.warning("Using deprecated runner %s. Consider upgrading to "
+                           "Beam 2.x." % self.runner)
+            return self.runner
+
+        else:
+            logger.warning("Found unsupported runner %s. Defaulting to "
+                           "DirectRunner." % self.runner)
+            return "DirectRunner"
+
+    def _get_dataflow_args(self):
+        def f(key, value):
+            return '--{}={}'.format(key, value)
+
+        output = []
+
+        runner = self._get_runner()
+        if runner:
+            output.append(f(self.dataflow_params.runner, runner))
```

Comment:
`runner` always has a value (defaults to `DirectRunner`). If a user forgets to overwrite `dataflow_params`, then this will fail with an NPE.

Seems like `dataflow_params` should be a required property, or have a default implementation as a fallback.

Reply:
If we don't default to `DirectRunner` then the default implementation for `DataflowParams` might not be needed.

Reply:
thx! in the `__init__` for `BeamDataflowTask` there's a check that `self.dataflow_params` is an instance of `DataflowParamKeys`. that should be enough of a check, right?

I modified the behavior so that if `self.runner` isn't specified, it defaults to `DirectRunner` with a warning. Otherwise, if any runner that isn't `DirectRunner`/`DataflowRunner` is specified, an exception is raised.